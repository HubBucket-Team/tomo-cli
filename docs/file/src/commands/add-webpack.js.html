<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../../">
  <title data-ice="title">src/commands/add-webpack.js | tomo-cli</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  <script src="script/manual.js"></script>
<meta name="description" content="A friendly command line tool designed to help create sustainable software using web technology"><meta property="twitter:card" content="summary"><meta property="twitter:title" content="tomo-cli"><meta property="twitter:description" content="A friendly command line tool designed to help create sustainable software using web technology"></head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./">Home</a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  <a href="test.html" data-ice="testLink">Test</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
<a style="position:relative; top:3px;" href="https://github.com/jhwohlgemuth/tomo-cli"><img width="20px" src="./image/github.png"></a></header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/ui.js~UI.html">UI</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-CommandError">CommandError</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-Debug">Debug</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-OfflineWarning">OfflineWarning</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-Task">Task</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-TaskList">TaskList</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-Warning">Warning</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-populateQueue">populateQueue</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#commands">commands</a><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addA11y">addA11y</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addBabel">addBabel</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addBrowsersync">addBrowsersync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addEsdoc">addEsdoc</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addJest">addJest</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addMakefile">addMakefile</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addPostcss">addPostcss</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-removePostcss">removePostcss</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-rustTasks">rustTasks</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-addWebpack">addWebpack</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-removeWebpack">removeWebpack</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-typedef">T</span><span data-ice="name"><span><a href="typedef/index.html#static-typedef-task">task</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#commands-add-eslint">commands/add-eslint</a><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-tasks">tasks</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#commands-add-marionette">commands/add-marionette</a><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-tasks">tasks</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#utils">utils</a><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/BasicEditor.js~BasicEditor.html">BasicEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/MakefileEditor.js~MakefileEditor.html">MakefileEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/Scaffolder.js~Scaffolder.html">Scaffolder</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoExist">allDoExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoExistSync">allDoExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoNotExist">allDoNotExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoNotExistSync">allDoNotExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-format">format</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getBinDirectory">getBinDirectory</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-parse">parse</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-someDoExist">someDoExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-someDoExistSync">someDoExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-createJsonEditor">createJsonEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-createModuleEditor">createModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getIntendedInput">getIntendedInput</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getVersions">getVersions</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-install">install</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-uninstall">uninstall</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-verifyRustInstallation">verifyRustInstallation</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-BabelConfigModuleEditor">BabelConfigModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-EslintConfigModuleEditor">EslintConfigModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-PackageJsonEditor">PackageJsonEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-PostcssConfigEditor">PostcssConfigEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-WebpackConfigEditor">WebpackConfigEditor</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">src/commands/add-webpack.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">import {
    PackageJsonEditor,
    WebpackConfigEditor,
    install,
    uninstall
} from &apos;../utils&apos;;
import {allDoExist, allDoExistSync, allDoNotExist, someDoExist} from &apos;../utils/common&apos;;

const WEBPACK_DEPENDENCIES = [
    &apos;cpy-cli&apos;,
    &apos;del-cli&apos;,
    &apos;npm-run-all&apos;,
    &apos;webpack&apos;,
    &apos;webpack-cli&apos;,
    &apos;webpack-dashboard&apos;,
    &apos;webpack-jarvis&apos;,
    &apos;webpack-dev-server&apos;,
    &apos;babel-loader&apos;
];
/**
 * @type {task[]}
 * @see https://webpack.js.org/
 */
export const addWebpack = [
    {
        text: &apos;Create Webpack configuration file&apos;,
        task: async ({sourceDirectory}) =&gt; {
            const entry = {
                app: `&apos;${sourceDirectory}/main.js&apos;`
            };
            const resolve = {
                modules: `[resolve(__dirname, &apos;${sourceDirectory}&apos;), &apos;node_modules&apos;]`
            };
            await (new WebpackConfigEditor())
                .create()
                .prepend(`const DashboardPlugin = require(&apos;webpack-dashboard/plugin&apos;);`)
                .prepend(`const {resolve} = require(&apos;path&apos;);`)
                .prepend(`/* eslint-env node */`)
                .extend({entry, resolve})
                .commit();
        },
        condition: () =&gt; allDoNotExist(&apos;webpack.config.js&apos;)
    },
    {
        text: &apos;Add build tasks to package.json&apos;,
        task: async ({outputDirectory, sourceDirectory}) =&gt; {
            const scripts = {
                copy: &apos;npm-run-all --parallel copy:assets copy:index&apos;,
                &apos;copy:assets&apos;: `cpy &apos;./assets/!(css)/**/*.*&apos; &apos;./assets/**/[.]*&apos; ${outputDirectory} --parents --recursive`,
                &apos;copy:index&apos;: `cpy &apos;./assets/index.html&apos; ${outputDirectory}`,
                prebuild: `del-cli ${outputDirectory}/assets`,
                build: &apos;webpack&apos;,
                postbuild: &apos;npm run copy&apos;,
                &apos;build:watch&apos;: `watch &apos;npm run build&apos; ${sourceDirectory}`,
                &apos;build:dashboard&apos;: &apos;webpack-dashboard -- webpack-dev-server --config ./webpack.config.js&apos;
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();

        },
        condition: () =&gt; someDoExist(&apos;package.json&apos;)
    },
    {
        text: &apos;Add dev task to package.json&apos;,
        task: async () =&gt; {
            const scripts = {
                dev: &apos;npm-run-all --parallel build:watch build:css:watch&apos;
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () =&gt; allDoExist(&apos;package.json&apos;, &apos;postcss.config.js&apos;),
        optional: () =&gt; allDoExistSync(&apos;package.json&apos;, &apos;postcss.config.js&apos;)
    },
    {
        text: &apos;Install Webpack dependencies&apos;,
        task: ({skipInstall}) =&gt; install(WEBPACK_DEPENDENCIES, {dev: true, skipInstall}),
        condition: ({isNotOffline}) =&gt; isNotOffline &amp;&amp; someDoExist(&apos;package.json&apos;)
    }
];
export const removeWebpack = [
    {
        text: &apos;Delete Webpack configuration file&apos;,
        task: async () =&gt; {
            await (new WebpackConfigEditor())
                .delete()
                .commit();
        },
        condition: () =&gt; someDoExist(&apos;webpack.config.js&apos;)
    },
    {
        text: &apos;Remove Webpack build tasks from package.json&apos;,
        task: async () =&gt; {
            const scripts = {
                copy: undefined,
                &apos;copy:assets&apos;: undefined,
                &apos;copy:index&apos;: undefined,
                prebuild: undefined,
                build: undefined,
                postbuild: undefined,
                &apos;build:watch&apos;: undefined,
                &apos;build:dashboard&apos;: undefined
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();

        },
        condition: () =&gt; someDoExist(&apos;package.json&apos;)
    },
    {
        text: &apos;Uninstall Webpack dependencies&apos;,
        task: () =&gt; uninstall(WEBPACK_DEPENDENCIES),
        condition: ({skipInstall}) =&gt; !skipInstall &amp;&amp; someDoExist(&apos;package.json&apos;) &amp;&amp; (new PackageJsonEditor()).hasAll(...WEBPACK_DEPENDENCIES),
        optional: ({skipInstall}) =&gt; !skipInstall
    }
];
export default addWebpack;</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(1.1.0)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
