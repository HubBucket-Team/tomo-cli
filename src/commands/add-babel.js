import {
    allDoNotExist,
    BabelConfigModuleEditor,
    install,
    someDoExist
} from '../utils';

const cfg = new BabelConfigModuleEditor();

const BABEL_DEPENDENCIES = [
    '@babel/cli',
    '@babel/core',
    '@babel/runtime'
];
const BABEL_PRESETS = [
    '@babel/preset-env'
];
const BABEL_PLUGINS = [
    '@babel/plugin-transform-runtime',
    '@babel/plugin-proposal-class-properties',
    '@babel/plugin-proposal-export-default-from',
    '@babel/plugin-proposal-optional-chaining'
];
const BABEL_REACT_PRESET = [
    '@babel/preset-react'
];
export default [
    {
        text: 'Create Babel config file',
        task: () => cfg.create(),
        condition: () => allDoNotExist('babel.config.js', '.babelrc', '.babelrc.js')
    },
    {
        text: 'Install Babel dependencies',
        task: ({skipInstall}) => install(BABEL_DEPENDENCIES, {dev: true, skipInstall}),
        condition: () => someDoExist('package.json')
    },
    {
        text: 'Install Babel presets',
        task: ({skipInstall}) => install(BABEL_PRESETS, {dev: true, skipInstall}),
        condition: () => someDoExist('package.json')
    },
    {
        text: 'Install Babel plugins',
        task: ({skipInstall}) => install(BABEL_PLUGINS, {dev: true, skipInstall}),
        condition: () => someDoExist('package.json')
    },
    {
        text: 'Install Babel React preset',
        task: ({skipInstall}) => install(BABEL_REACT_PRESET, {dev: true, skipInstall}),
        condition: ({useReact}) => (useReact && someDoExist('package.json')),
        optional: ({useReact}) => useReact
    },
    {
        text: 'Add React support to Babel configuration file',
        task: () => cfg.extend({
            presets: [...BABEL_PRESETS, ...BABEL_REACT_PRESET]
        }),
        condition: ({useReact}) => (useReact && someDoExist('babel.config.js')),
        optional: ({useReact}) => useReact
    }
];