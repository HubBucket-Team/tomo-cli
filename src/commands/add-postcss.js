import {
    allDoNotExist,
    install,
    PackageJsonEditor,
    PostcssConfigEditor,
    someDoExist
} from '../utils';

const pkg = new PackageJsonEditor();
const cfg = new PostcssConfigEditor();

const POSTCSS_DEPENDENCIES = [
    'cssnano',
    'postcss-cli',
    'postcss-reporter',
    'postcss-safe-parser',
    'postcss-import',
    'postcss-cssnext',
    'stylelint',
    'uncss'
];

export default [
    {
        text: 'Create PostCSS config file',
        task: () => cfg.create(),
        condition: () => allDoNotExist('postcss.config.js')
    },
    {
        text: 'Install PostCSS dependencies',
        task: ({skipInstall}) => install([POSTCSS_DEPENDENCIES], {dev: true, skipInstall}),
        condition: () => someDoExist('package.json')
    },
    {
        text: 'Add PostCSS tasks to package.json',
        task: () => pkg.extend({

        }),
        condition: () => someDoExist('package.json')
    }
];