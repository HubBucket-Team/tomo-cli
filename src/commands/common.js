import {mkdirp} from 'fs-extra';
import {
    allDoNotExist,
    PackageJsonEditor
} from '../utils';

export const createPackageJson = [
    {
        text: 'Create package.json',
        task: () => {
            const pkg = new PackageJsonEditor();
            pkg.create();
        },
        condition: () => allDoNotExist('package.json')
    }
];
export const createSourceDirectory = [
    {
        text: 'Create source directory',
        task: ({sourceDirectory}) => mkdirp(sourceDirectory),
        condition: ({sourceDirectory}) => allDoNotExist(sourceDirectory)
    }
];