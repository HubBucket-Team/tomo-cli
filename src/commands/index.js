/* eslint-disable no-magic-numbers */
import {
    createPackageJson,
    createSourceDirectory
} from './common';
import addA11y from './add-a11y';
import addBabel from './add-babel';
import addEsdoc from './add-esdoc';
import addEslint from './add-eslint';
import addJest from './add-jest';
import addMarionette from './add-marionette';
import addPostcss from './add-postcss';
import addWebpack from './add-webpack';

const createProject = [
    ...createPackageJson,
    ...createSourceDirectory,
    ...addBabel,
    ...addEslint,
    ...addJest
];
const create = {
    project: createProject,
    app: [
        ...createProject,
        ...addMarionette
    ],
    server: [
        ...createProject
    ]
};
const add = {
    a11y: addA11y,
    babel: addBabel,
    docs: addEsdoc,
    eslint: [
        ...addBabel,
        ...addEslint
    ],
    jest: [
        ...addBabel,
        ...addJest
    ],
    postcss: addPostcss,
    webpack: addWebpack
};

module.exports = {
    add,
    create,
    new: create// alias for create
};