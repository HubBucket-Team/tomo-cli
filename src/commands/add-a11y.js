import {
    PackageJsonEditor,
    install,
    uninstall
} from '../utils';
import {allDoExist} from '../utils/common';
/**
 * @type {task[]}
 * @see http://pa11y.org/
 */
export const addA11y = [
    {
        text: 'Add pa11y tasks to package.json',
        task: async ({outputDirectory}) => {
            const scripts = {
                'lint:a11y': `pa11y ${outputDirectory}/index.html`
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Install pa11y',
        task: ({skipInstall}) => install(['pa11y'], {dev: true, skipInstall}),
        condition: ({isNotOffline}) => isNotOffline && allDoExist('package.json')
    }
];
export const removeA11y = [
    {
        text: 'Remove pa11y tasks from package.json',
        task: async () => {
            const scripts = {
                'lint:a11y': undefined
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Uninstall pa11y',
        task: () => uninstall(['pa11y']),
        condition: ({skipInstall}) => !skipInstall && allDoExist('package.json') && (new PackageJsonEditor()).hasAll('pa11y'),
        optional: ({skipInstall}) => !skipInstall
    }
];
export default addA11y;