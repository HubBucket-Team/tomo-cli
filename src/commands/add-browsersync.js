import {PackageJsonEditor, install} from '../utils';
import {allDoExist, someDoExist} from '../utils/common';

const BROWSERSYNC_DEPENDENCIES = [
    'browser-sync',
    'npm-run-all'
];
/**
 * @type {task[]}
 * @see https://www.browsersync.io/docs/command-line
 */
export const addBrowsersync = [
    {
        text: 'Add Browsersync tasks to package.json',
        task: async ({outputDirectory}) => {
            const scripts = {
                prestart: 'npm run build',
                start: 'npm-run-all --parallel build:watch build:css:watch serve:dev',
                'serve:dev': `browser-sync start --server ${outputDirectory} --files ${outputDirectory}`
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json', 'webpack.config.js', 'postcss.config.js')
    },
    {
        text: 'Install Browsersync dependencies',
        task: ({skipInstall}) => install(BROWSERSYNC_DEPENDENCIES, {dev: true, skipInstall}),
        condition: ({isNotOffline}) => isNotOffline && someDoExist('package.json')
    }
];