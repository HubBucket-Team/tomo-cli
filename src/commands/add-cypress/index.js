import {join} from 'path';
import {EslintConfigModuleEditor, PackageJsonEditor, install, uninstall} from '../../utils';
import {allDoExist, allDoNotExist} from '../../utils/common';
import {Scaffolder} from '../../utils/Scaffolder';
import createJsonEditor from '../../utils/createJsonEditor';

const ALWAYS = () => true;
const CYPRESS_DEPENDENCIES = [
    'cypress',
    'cypress-match-screenshot'
];
/**
 * @type {task[]}
 * @see https://www.cypress.io/
 */
export const addCypress = [
    {
        text: 'Create Cypress config file',
        task: async ({port}) => {
            const Editor = createJsonEditor('cypress.json', {
                baseUrl: `http://localhost:${port}`
            });
            await (new Editor())
                .create()
                .commit();
        },
        condition: () => allDoNotExist('cypress.json')
    },
    {
        text: 'Add Cypress test tasks to package.json',
        task: async () => {
            const scripts = {
                cypress: 'cypress open',
                'test:e2e': 'npm-run-all --parallel start cypress'
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Add Cypress global variable to .eslintrc.js',
        task: async () => {
            const globals = {cy: true};
            await (new EslintConfigModuleEditor())
                .extend({globals})
                .commit();
        },
        condition: () => allDoExist('.eslintrc.js')
    },
    {
        text: 'Copy Cypress files',
        task: async ({overwrite}) => {
            await (new Scaffolder(join(__dirname, 'templates')))
                .overwrite(overwrite)
                .target('cypress/support')
                .copy('support/index.js', 'index.js')
                .copy('support/commands.js', 'commands.js')
                .target('cypress/integration')
                .copy('visual-regression.test.js')
                .commit();
        },
        condition: ALWAYS
    },
    {
        text: 'Install Cypress dependencies',
        task: ({skipInstall}) => install([...CYPRESS_DEPENDENCIES, 'npm-run-all'], {dev: true, skipInstall}),
        condition: ({isNotOffline}) => isNotOffline && allDoExist('package.json')
    }
];
export const removeCypress = [
    {
        text: 'Delete Cypress config file',
        task: async () => {
            const Editor = createJsonEditor('cypress.json');
            await (new Editor())
                .delete()
                .commit();
        },
        condition: () => allDoExist('cypress.json')
    },
    {
        text: 'Remove Cypress test tasks from package.json',
        task: async () => {
            const scripts = {
                cypress: undefined,
                'test:e2e': undefined
            };
            await (new PackageJsonEditor())
                .extend({scripts})
                .commit();
        },
        condition: () => allDoExist('package.json')
    },
    {
        text: 'Remove Cypress global variable from .eslintrc.js',
        task: async () => {
            const globals = {cy: false};
            await (new EslintConfigModuleEditor())
                .extend({globals})
                .commit();
        },
        condition: () => allDoExist('.eslintrc.js')
    },
    {
        text: 'Uninstall Cypress dependencies',
        task: () => uninstall(CYPRESS_DEPENDENCIES),
        condition: ({skipInstall}) => !skipInstall && allDoExist('package.json') && (new PackageJsonEditor()).hasAll(...CYPRESS_DEPENDENCIES),
        optional: ({skipInstall}) => !skipInstall
    }
];
export default addCypress;