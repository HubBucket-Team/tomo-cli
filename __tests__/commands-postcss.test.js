import {fileContents, getDirectoryTree, run, useTemporaryDirectory} from './tomo-test';
import {createPackageJson} from '../src/commands/common';
import {addPostcss, removePostcss} from '../src/commands/add-postcss';

jest.mock('is-online', () => (async () => true));

describe('PostCSS', () => {
    let tempDirectory;
    const skipInstall = true;
    const assetsDirectory = './assets';
    const outputDirectory = './dist';
    const options = {assetsDirectory, skipInstall, outputDirectory};
    const omit = ['extension', 'path', 'size', 'type'];
    const [setTempDir, cleanupTempDir] = useTemporaryDirectory();
    beforeEach(async () => {
        tempDirectory = await setTempDir();
        process.chdir(tempDirectory);
    });
    afterEach(async () => {
        await cleanupTempDir();
    });
    test('Add support', async () => {
        await run(createPackageJson, {});
        await run(addPostcss, options);
        const tree = getDirectoryTree(tempDirectory, {omit});
        const packageJson = fileContents('./package.json');
        const postcssConfig = fileContents('./postcss.config.js');
        expect(tree).toMatchSnapshot();
        expect(packageJson).toMatchSnapshot();
        expect(postcssConfig).toMatchSnapshot();
    });
    test('Remove support', async () => {
        await run(createPackageJson, {});
        await run(addPostcss, options);
        await run(removePostcss, {skipInstall});
        const tree = getDirectoryTree(tempDirectory, {omit});
        const packageJson = fileContents('./package.json');
        expect(tree).toMatchSnapshot();
        expect(packageJson).toMatchSnapshot();
    });
});