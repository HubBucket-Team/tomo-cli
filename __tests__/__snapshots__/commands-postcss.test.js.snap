// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCSS Add support 1`] = `
"{
    name: tomo-test,
    children: [
        {
            name: package.json
        },
        {
            name: postcss.config.js
        }
    ]
}
"
`;

exports[`PostCSS Add support 2`] = `
"{
    \\"name\\": \\"my-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"description\\": \\"A super cool app/server/tool/library/widget/thingy\\",
    \\"license\\": \\"MIT\\",
    \\"keywords\\": [],
    \\"scripts\\": {
        \\"build:css\\": \\"postcss ./assets/css/style.css --dir ./dist\\",
        \\"watch:css\\": \\"npm run build:css -- --watch\\"
    }
}
"
`;

exports[`PostCSS Add support 3`] = `
"module.exports = {
    map: true,
    parser: require('postcss-safe-parser'),
    plugins: [
        require('stylelint')({config: {extends: 'stylelint-config-recommended'}}),
        require('postcss-import')(),
        require('postcss-preset-env')({stage: 0}),
        require('cssnano')(),
        require('postcss-reporter')({clearReportedMessages: true})
    ]
};"
`;

exports[`PostCSS Remove support 1`] = `
"{
    name: tomo-test,
    children: [
        {
            name: package.json
        }
    ]
}
"
`;

exports[`PostCSS Remove support 2`] = `
"{
    \\"name\\": \\"my-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"description\\": \\"A super cool app/server/tool/library/widget/thingy\\",
    \\"license\\": \\"MIT\\",
    \\"keywords\\": [],
    \\"scripts\\": {}
}
"
`;
